File: exercise/php/day24/tests/Doubles/InMemoryToyRepository.php
<?php

namespace Tests\Doubles;

use Delivery\Domain\IToyRepository;
use Delivery\Domain\Toy;

class InMemoryToyRepository implements IToyRepository
{
    private array $toys = [];
    private array $raisedEvents = [];

    public function findByName(string $toyName): ?Toy
    {
        foreach ($this->toys as $toy) {
            if ($toy->getName() === $toyName) {
                return $toy;
            }
        }
        return null;
    }

    public function findById(string $id): ?Toy
    {
        return $this->toys[$id] ?? null;
    }

    public function save(Toy $toy): void
    {
        $this->raisedEvents = [];
        $this->toys[$toy->getId()] = $toy;

        foreach ($toy->getUncommittedEvents() as $event) {
            $this->raisedEvents[] = $event;
        }

        $toy->clearUncommittedEvents();
    }

    public function raisedEvents(): array
    {
        return $this->raisedEvents;
    }
}

File: exercise/java/day24/src/test/java/delivery/doubles/InMemoryToyRepository.java
package delivery.doubles;

import domain.Toy;
import domain.ToyRepository;
import domain.core.Aggregate;
import domain.core.Event;
import io.vavr.collection.HashMap;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import io.vavr.collection.Seq;
import io.vavr.control.Option;

import java.util.UUID;

public class InMemoryToyRepository implements ToyRepository {
    private Map<UUID, Toy> toys = HashMap.empty();
    private Seq<Event> raisedEvents = List.empty();

    @Override
    public Option<Toy> findByName(String toyName) {
        return toys.filter(toy -> toy._2().getName().equals(toyName))
                .values()
                .headOption();
    }

    @Override
    public Option<Toy> findById(UUID id) {
        return toys.get(id);
    }

    @Override
    public void save(Toy toy) {
        raisedEvents = List.empty();
        toys = toys.put(toy.getId(), toy);

        ((Aggregate) toy).getUncommittedEvents()
                .forEach(event -> raisedEvents = raisedEvents.append(event));

        ((Aggregate) toy).clearUncommittedEvents();
    }

    public Seq<Event> raisedEvents() {
        return raisedEvents;
    }
}